### ISSUE TITLE
Returning Retry on ExecuteFilesInUse causes hang on MsiExec

### ISSUE THREAD

**Issue by Danielku15** [wixtoolset/issues#5790]
* Which version of WiX are you building with?

> 3.11.1.2318

* Which version of Visual Studio are you building with (if any)?

> Visual Studio 2017 Professional (15.5.2)

* Which version of the WiX Toolset Visual Studio Extension are you building with (if any)?

> 0.9.21.62588

* Which version of .NET are you building with?

> .NET 4.7.1

* If the problem occurs when installing your packages built with WiX, what is the version of Windows the package is running on?

> Windows 10 Professional (1709)

* Describe the problem and the steps to reproduce it.

In my installer I tried to cover the case of files being used by applications by showing the typical dialog that files are in use by subscribing the `ExecuteFilesInUse` event. But unlike the typical selection of forcly close the applications or trigger a reboot, I want to give the user a chance to manually close the application. That's why I made a dialog with a Retry/Cancel option. This works fine from a UI perspective. But when you close all applications that use files, and the installation continues, the msiexec gets stuck in the uninstall sequence. 

Steps to reproduce are: 

* [Code] Subscribe to the `ExecuteFilesInUse` event. 
* [Code] Set the `Result` property to `Retry` in case files are locked. 
* [Action] Install the product
* [Action] Lock any of the files placed by your installer (e.g. via command line `notepad >FileInInstallDir`)
* [Action] Uninstall the product
* [Action] Click on retry when the retry popup appears
* [Action] Close the notepad
* [Action] Click on retry

Result: msiexec.exe gets stuck. No additional logs are produced which makes it unclear to me why the sequence gets stuck but it looks like an endless loop to me as the msiexec.exe is utilizing 1 CPU core at 100%

Here is a full example of such an installer which allows reproduction of the above steps easily: 
[WixBootstrapper.zip](https://github.com/wixtoolset/issues/files/1870727/WixBootstrapper.zip)

* Describe the behavior you expected and how it differed from the actual behavior.

After the file lock is resolved, and the installer continues, I expect the `msiexec.exe` to finish the uninstallation by removing the files. 

> Note: I also filed a question on [StackOverflow](https://stackoverflow.com/questions/49517794/handling-executefilesinuse-with-retry-and-resolving-used-files) last week but there was no response. As I was able to reproduce this issue also in a fresh standalone example I decided to file a issue here as it looks like a clear malfunction to me.

**Comment by barnson** [comment]
Not clear this is a WiX bug. Can you get a dump showing Burn in the hung thread?

**Comment by Danielku15** [comment]
Is there maybe any further information on this topic? Our iteration is about to end and this appears to be the last remaining task for our installer. At least some feedback would be nice.

**Comment by barnson** [comment]
This doesn't appear to be a WiX bug.

**Comment by Danielku15** [comment]
What do mean exactly by this doesn't appear to be a WiX bug? I'd be grateful for some details. 

- Do you mean it is not a **bug** and an expected behavior? If so, why would this be this way? 
- Do you mean it is not a **WiX** bug and caused by something else? If so, you have already a guess why and how this can happen? 

Is there a chance that somebody from the WiX team will give this a look to really proof where this problem comes from?

For me it sounds very strange that just because a file was locked, the uninstall sequence suddenly requires multiple minutes. I haven't traced back on how this event is provided by WiX but I guess it is an event that comes from MSI? In our old (non WiX) installer this feature appeared to work fine but I guess they used another system to detect "files in use" as they showed the locked files, not the processes using the files. 

I also guess WiX does not provide any other system to check for files in use?

**Comment by BMurri** [comment]
Burn uses the MsiSetExternalUIRecord API (if it exists, which it does on all supported versions of Windows today) or MsiSetExternalUI (if it doesn't). In either case it uses the Unicode version of the API (e.g. MsiSetExternalUIRecordW). Either MSI or any custom action reporting files-in-use to MSI will report through MSI to Burn using that API, and Burn passes it along to your BA.

I could try guessing as to why MSI/restart manager might be taking so long, but the system event log might give you more information. Either way, I also don't see a WiX bug here either.

Without seeing the entire stack dump, I'd conjecture that the win32u.dll thread is msi.dll calling the service side MsiExec.exe process, but there's no way for me to know for sure without the dump.

**Comment by barnson** [comment]
Agree with @BMurri that it's MSI or MSI waiting on Restart Manager.

**Comment by barnson** [comment]
Per triage, this could be in MSI, Restart Manager, your custom BA, or Burn. It's going to take fairly deep debugging to narrow down the cause.

**Comment by Danielku15** [comment]
We for not decided to completely remove the FilesInUse feature. It is definitly too buggy/non-deterministic for our needs. I'm not sure how the restart manager or MSI exactly detect the files in use, but we very often get false notifications. Just 2 examples:

1) With the event enabled it is highlighted that various processes are using files. But in the same scenario, when disabling the FileInUse, installation works smooth (no other error/warning events are triggered and also no PengingFileRenameOperations are registered). 

2) We have 2 of our products using the same DLL. The products come with separate installers and are located in separate installation folders. When uninstalling one application it highlights files in use by the other application even though it is definitly not using anything from the first application. The DLL has the same footprint and maybe Windows does some special code-segment sharing across the applications but it is not acceptable for us. 

Even though the issue is not really solved, it would be fine for me to close this issue here as we will not rely on this feature anymore. Here some notes for other devs who might run into similar issues: 

- Golden Rule: Try to avoid the FilesInUse event handling.
- The data format is not consistent and can change depending on the source. 
- The event lists only applications using files, not the files in use itself. 
- The event often reports wrong information. 
- Hooking into the event with retry operations can lead to hangs on the uninstall/upgrade sequence. If you're lucky it might resolve itself after a long waiting time. 
- Personal recommendation as solution (might be architectural-wise not perfect): With best knowledge about your application try to close/shutdown all processes/services of your application. Best is if you do this in a special MSI which is not responsible for any file copy operations. Place this MSI at the beginning of your installer chain. This way components (RestartManager/MSI) which might check if files are in use, will not detect anything anymore as applications are already closed. Depending on the phase those components do their checks you might not be able to close applications within the same MSI shipping the product. Especially if you might need to do elevated/deferred custom actions to kill your processes (e.g. if you have tray icons reacting to a WM_CLOSE with minimizing).

**Comment by BMurri** [comment]
During some of my previous experiences with older versions of MSI (we're talking over 15 years now) I did some tests that showed various FilesInUse findings to help myself understand its behavior. Some of those findings include the following:
* Processes, rather than the files themselves, are shown because that's user actionable information. MSI doesn't pretend to be smart enough to know how the process is making use of the file. Further, this message is intended for files that either MSI directly or Restart Manager aren't already prepared to "automagically" handle.
* Files held by processes that had been created by system services that the MSI itself controlled (and those processes were managed by the service, such that a clean controlled shutdown as per a correctly authored MSI) never produced a reboot required scenario, but under some conditions produced reports (especially if the service communicated with "grandchildren" processes and the intermediate processes had exited).
* Files with the same names as the ones you are replacing/removing would sometimes trigger the reports (which you note above. It seems that some part of the code inside of either MSI or Restart Manager (I never determined which) didn't always seem to take into account file paths).
* I've seen MSI "lock up" like you describe when a custom action unexpectedly crashed, leaving no trace of why, except that it didn't spin a single CPU to 100% (everything was waiting for some event that never materialized). MSI itself would usually stop with an error after approx. 21 minutes, though that didn't always happen either. Reproducing the crash outside of MSI took months, after accidentally finding out the fact of the crash.
* In Windows, there's (at least) 3-4 different ways to "lock" a file, each one with different effects on the various tools MSI employs for both conflict detection and reboot mitigation.

MSI itself hasn't changed an awful lot in the past 15 years, so I generally look back at my list when I see something "strange".

FilesInUse reports are based on a snapshot in time, before any side effects of the installation transaction have been performed, and are expected to be advisory at best. Further, it doesn't protect you from some process or system accessing files after it runs before the files are replaced/removed.

Knowing something about your product and how your installation relates to your running processes allows you to intelligently evaluate whether some part of the report is potentially expected or not. In the most naïve case, stopping everything found will eliminate almost all reboots ultimately caused by files-in-use, but at a cost of stopping more than the expected amount, which given the degree to which some customers hate reboots can seem a small price to pay.

If all found processes are written to interface properly with Restart Manager, the only hard part is restarting elevated processes, in which case you are on your own. However, if you are maintaining code that doesn't follow the documented [protocols](https://msdn.microsoft.com/library/windows/desktop/aa373652.aspx) defined by Restart Manager (hint, WM_ENDSESSION w/ lParam of ENDSESSION_CLOSEAPP is sent **before** WM_CLOSE) then yes, you have to do all sorts of gymnastics to get to 99.9% reboot avoidance, and those applications are hard on everybody.

A minidump of msiexec without debug symbols and source code for (pick at least one of) either your code, our code, or Microsoft's code, isn't going to be productive, and unless you have evidence that one of our custom actions may be the one either locking up or crashing (even if it's within the next 2-3 actions after the last one showing up in the log) the fishing expedition inside of those minidumps may only be successful for someone with access to Windows Sources (and there are very few of those people actively working on the WiX toolset). This isn't a common occurrence.

Message [2727](https://msdn.microsoft.com/library/windows/desktop/aa372835.aspx) is "The directory entry '[2]' does not exist in the Directory table." [InstallValidate](https://msdn.microsoft.com/library/windows/desktop/aa369546.aspx) is the built-in action that runs the UI portion of FilesInUse, including both the discovery phase of the interaction between Restart Manager and MSI, and sending the FilesInUse messages out (ultimately to the BA). I don't know if that could be related to your hang or not (and there's still insufficient information here to tell).

If you add a UI to your MSI, including an MsiRMFilesInUse dialog (WiX provides one), can you reproduce this (leaving Burn out of the picture entirely)? If so, sending the MSI and repro instructions to Microsoft may allow them to determine if the bug is on their side. If not, we would be interested in knowing about it, because that would seem to narrow down the scope of possible environmental factors to potentially include some of our code.

**Comment by rseanhall** [comment]
I can't reproduce a hang using the example provided. There is a few seconds of delay, which probably comes from the documented behavior of [redoing costing](https://docs.microsoft.com/en-us/windows/win32/msi/installvalidate-action) on retry.

The BA is automatically sending `Retry` if any of the files given in `OnFilesInUse` is null or empty. That could theoretically cause an infinite loop.

The example BA is also calling `Apply` with `IntPtr.Zero`. We have seen this cause MSI to hang in a different scenario. The WixBA gets the WPF window handle like this:

https://github.com/wixtoolset/wix3/blob/d44e926a00599534593ada3239884456bb2455a8/src/Setup/WixBA/RootView.xaml.cs#L28
